File: ScenarioApproach/myFunctions/GetEqualityConstraints.m
function [ABig, bBig] = GetEqualityConstraints(G, ODs)

% extract dimensions
numNode = G.numnodes;
numEdge = G.numedges;
numDmnd = size(ODs, 1);

% use memory-saving sparse matrix
A = sparse(numNode, numEdge);
b = sparse(numNode, numDmnd);

% construct matrix A
for n = 1:numNode
    edgeIn  =  inedges(G, n); % row index in edge table
    edgeOut = outedges(G, n); % row index in edge table
    
    A(n, edgeIn)  =  1; % sum of inedge flows
    A(n, edgeOut) = -1; % sum of outedge flows
end

% construct matrix b
for k = 1:numDmnd
    s = ODs(k, 1);
    t = ODs(k, 2);
    d = ODs(k, 3);
    
    b(s, k) = -d;
    b(t, k) =  d;
end

% augmented matrix A (x stacked into vector)
A11                        =  -eye(numEdge); %        |     |         |
A12      = repmat(eye(numEdge), 1, numDmnd); %        | A11 |   A12   | 
A21      = sparse(numNode*numDmnd, numEdge); %        |-----+---------|
ACell             = repmat({A}, numDmnd, 1); % ABig = |     |         |     
A22                     = blkdiag(ACell{:}); %        | A21 |   A22   |       
ABig                 = [A11, A12; A21, A22]; %        |     |         |

% augmented vector b (x stacked into vector)
bBigDown = reshape(b, [numNode*numDmnd, 1]); % bBD  = [b1; ... ; bK]
bBig       = [sparse(numEdge, 1); bBigDown]; % bBig = [0; bBD];

end

File: ScenarioApproach/myPython/GetEqualityConstraints.py
import numpy as np
from numpy import matlib


def GetEqualityConstraints(G, ODs):

    M = G.numEdge # number of edges
    N = G.numNode # number of nodes
    K = G.numDmnd # number of demands

    # compute net in-flow of each node
    A = np.zeros((N, M), dtype=np.double)
    for idxLink, link in enumerate(G.dataNet):
        idxInit = int(link[0]) - 1 # index starts from 0
        idxTerm = int(link[1]) - 1 # index starts from 0
        A[idxInit][idxLink] = A[idxInit][idxLink] - 1
        A[idxTerm][idxLink] = A[idxTerm][idxLink] + 1

    A11 = -np.eye(M, dtype=np.double)
    A12 = np.matlib.repmat(np.eye(M, dtype=np.double), 1, K)
    A21 = np.zeros((N*K, M), dtype=np.double)
    A22 = np.kron(np.eye(K, dtype=np.double), A) # arrange matrix A along diagonal

    A1   = np.concatenate((A11, A12), axis=1) #        | A11 | A12 |   | A1 |
    A2   = np.concatenate((A21, A22), axis=1) # ABig = |-----+-----| = |----|
    ABig = np.concatenate((A1,  A2 ), axis=0) #        | A21 | A22 |   | A2 |

    b = np.zeros((N, K), dtype=np.double)
    for k in range(K):
        idxInit = int(ODs[k][0]) - 1 # index starts from 0
        idxTerm = int(ODs[k][1]) - 1 # index starts from 0
        demand  = ODs[k][2]

        b[idxInit][k] = -demand # net out-flows equal demand
        b[idxTerm][k] =  demand # net in-flows equal demand
    
    bBigLow = np.reshape(b, (N*K, 1), order='F')        #        |    0    |
    bBigUp  = np.zeros((M, 1), dtype=np.double)         # bBig = |---------|
    bBig    = np.concatenate((bBigUp, bBigLow), axis=0) #        | bBigLow |

    return [ABig, bBig]